### ISSUE TITLE
Nested arrays parsed incorrectly

### ISSUE THREAD

**Issue by oliversturm** [ljharb/qs#212]
Hi,

I have this object:

```js
{
  filter: [
    [ 'int1', '=', 77 ],
    'and',
    [ 'int2', '=', 8 ]
  ]
}
```

Using jQuery `$.param()` (or jquery-param), this gives me the following query string:

```
filter%5B0%5D%5B%5D=int1&filter%5B0%5D%5B%5D=%3D&filter%5B0%5D%5B%5D=77&filter%5B%5D=and&filter%5B2%5D%5B%5D=int2&filter%5B2%5D%5B%5D=%3D&filter%5B2%5D%5B%5D=8

// decoded:
filter[0][]=int1&filter[0][]==&filter[0][]=77&filter[]=and&filter[2][]=int2&filter[2][]==&filter[2][]=8
```

Unfortunately, `qs.parse()` returns this instead of the original object:

```js
{
  filter: [ 
    [ 'int1', '=', '77', 'and' ], 
    [ 'int2', '=', '8' ] 
  ] 
}
```

The problem seems to be the part `filter[]=and` -- if I use `filter[1]=and` instead, the result is correct. However, the syntax makes sense to me the way it is, so I'm wondering if this is a bug in qs. Any ideas?

Thanks
Oli

**Comment by ljharb** [comment]
What happens if you pass the `{ arrayFormat: 'indices' }` option to `qs.parse`? The default is `brackets`.

**Comment by oliversturm** [comment]
@ljharb Unfortunately this doesn't seem to make any difference - I get the same result with or without any `arrayFormat`.

**Comment by ljharb** [comment]
ahhhh yes.

If you use indices in an array, you have to use them everywhere - it's not like code where you can just stick a random `[]` in there and it acts like "push".

So, I think that either it all has to use `[]`, or, it all has to use indices.

**Comment by oliversturm** [comment]
@ljharb Well, like I said it's not me doing this - jQuery `.param()` does it that way. And while I agree that it would look cleaner to have indices everywhere, the intention of the query string still seems clear and unambiguous to me. I can't judge whether this query string should be consider correct or not...

**Comment by ljharb** [comment]
Any reason not to use qs.stringify?

**Comment by oliversturm** [comment]
@ljharb Good question, it might be possible and I'll check it out. I guess it would still be good if qs understood the output from jQuery correctly ;)

**Comment by oliversturm** [comment]
Btw, I find that `{arrayFormat: 'indices'}` seems to be the default behavior for `stringify`? That's fine with me, just surprising if the default for `parse` is `brackets`. I also find (just out of curiosity) that results are surprising when using `brackets`:

```js
> qs.parse(qs.stringify({ filter: [ [ 'int1', '=', '77' ], 'and', [ 'int2', '=', '8' ] ] }, {arrayFormat: 'brackets'}), {arrayFormat: 'brackets'})
{ filter: [ 'int1', '=', '77', 'int2', '=', '8', 'and' ] }
```

Just FYI - I'm fine with the `indices` format and it works correctly. I'll see what it takes to use that instead of jQuery format for my purposes.

**Comment by ljharb** [comment]
This is still looking into regardless; I do agree with you that it should "just work".

Separately, if the default arrayFormat is indeed different between parse and stringify, that's something that needs fixing :-)

In general I think the weirdness is caused by having an array of "sometimes arrays and sometimes a string" - if you put "and" in an array, my hunch is that the weirdness would vanish.

**Comment by ljharb** [comment]
* ah - `parse` doesn't take `arrayFormat` at all :-) that's just a stringify option.

**Comment by oliversturm** [comment]
@ljharb I see! I was just going with your suggestion - had a hard time before finding a list of all the options somewhere, so I didn't check.

**Comment by ljharb** [comment]
That's my bad :-)

**Comment by jchook** [comment]
If someone could check my commit, I think it fixes the issue: 59a53c40ba9888df1ae86cd9ddacb08bc13e6bf1

**Comment by ljharb** [comment]
@jchook would you be able to make a PR, and we can explore it further there?
